ENH: adding simulated dual annealing to optimize

sgubianpm opened the pull request:
Dear All,

Old related PRs:
#6569
#6197

Thanks a lot for the time you spent on reviewing these PRs that you did quite long time ago now. Last @rgommers comment about adding a needs-work yellow label and @js850 commented that refactoring was needed as well. So I did the recent weeks.

First of all, the code has been refactored and cut into smaller classes pieces so that it gained a lot in readability while addressing as much as possible all of your comments.

Secondly, in the meantime, the method has evolved internally and relies now on Simulated Dual Annealing (and therefore has been renamed into SDA). It is showing even better results than the previous implementation in terms of successful rate and number of objective function calls.

In addition, as suggested by @js850, the possibility to use any of the SciPy minimizer to operate the local search has been added.

Maybe this changes could go towards a foreseeable merge.

Best regards,

Sylvain.

As mentioned by @rgommers, I haven't attached the  benchmarks results in this PR (thanks for reminding me). So, below is the summary for the benchmarks.
The benchmark has been done by using the SciPy global optimization testing functions implemented by SciPy contributors plus some of these functions using bigger dimensions (from 2 to 100 components). The SDA method has been tested against methods available in SciPy but also other Python available global optimizers like the [NLOpt](https://nlopt.readthedocs.io/en/latest/NLopt_Python_Reference/) suite, [PSO](https://pythonhosted.org/pyswarm/) (particle swarm) and [CMAES](https://github.com/CMA-ES/pycma) which is the is the Covariance Matrix Adaptation Evolution Strategy implementation. A cluster infrastructure has been used to execute 100 runs with different random initial points for the 250 testing functions over the 14 tested methods.
(The internal SciPy benchmark has not been used as it not meant to be run on a cluster, it does use the default testing function dimensions and it is dedicated to SciPy global optimization methods). 
Below figure are showing the overall results of the benchmark.
ESCH, MLSL, ISRES and CRS are NLOpt implementations, TSA stands for traditional simulated annealing. BF, BH and DE are the available SciPy methods (brute_force, basin_hopping, differential evolution). 

![sda_bench](https://user-images.githubusercontent.com/17434537/33834032-e39ebd9e-de81-11e7-9138-c1ab4d61376d.png)

(ANFC: Average number of objective function call used to reach the global optimum value (with a 1.e-8 absolute tolerance value).

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,589 @@
+        return self.func(x, *self.fun_args)
+
+    def gradient(self, x):
+        g = np.zeros(x.size, np.float64)
```
I think it would be useful to allow the user to have more freedom specifying the local search method parameters.

It would also be nice to be able to include the gradient of the local search method as an optional argume

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,589 @@
+def sda(func, x0, bounds, maxiter=1000, minimizer_kwargs=None,
+        initial_temp=5230., visit=2.62, accept=-5.0, maxfun=1e7, seed=None,
+        pure_sa=False):
+    """
```
I think it would be useful to allow the user to have more freedom specifying the local search method parameters. And I think you should also include the gradient of the local search method as an optional argument, and use finite-difference option only when this argument is not provided.

sgubianpm replied:
I thought that this was I did. Local search method, related arguments and gradient can be passed in "minimizer_kwargs" argument like for basin_hopping, if not specified then "L-BFGS-B" is used.Is this not correct?

antonior92 replied:
You are right. I have missed this out. Your implementation does gives this freedom to the user.

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,589 @@
+        else:
+            self.fun_args = ()
+        if not self.kwargs or 'method' not in self.kwargs:
+            self.kwargs['method'] = 'L-BFGS-B'
```
I think the choice of the local search method deserve some thought. 

By what I know of "L-BFGS-B", the method is very good for large scale problems, but since global optimization algorithms usually cannot handle problems with too many variable (>100 variables) very well, this may not be the best choice of local search method.

What are your thoughts on this?

sgubianpm replied:
Hi @antonior92, thanks a lot for your review and congratulation for being in the SciPy core team, this is great!
Regarding the choice of using "L-BFGS-B" as default local search method, in the context of the extensive benchmarks performed (see details in the PR description) this is the one that has proved the best efficiency/reliability even for small dimension problems.

antonior92 replied:
Oh, thank you very much for including the benchmarks. Your method does have some impressive results.

This choice did not matched my first intuition on the problem, but it puts my mind at easy that you have tested different methods and choose "L-BFGS-B" based on these results.

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,589 @@
+        self.nb_fun_call += 1
+        return self.func(x, *self.fun_args)
+
+    def gradient(self, x):
```
It seems to me that you are implementing another finite difference approximation. I would strongly recommend you to use the already implemented one at ``optimize/_numdiff.py``. You would avoid duplicating code. Furthermore, the implementation at ``_numdiff.py`` is carefully implemented and well tested.

sgubianpm replied:
Thanks for this, I have replaced the default gradient computation by the SciPy approx_derivative one.

antonior92 replied:
Great!

rgommers commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,589 @@
+    The following example is a 10-dimensional problem, with many local minima.
+    The function involved is called Rastrigin
+    (https://en.wikipedia.org/wiki/Rastrigin_function)
+    >>> import numpy as np
```
This import should be removed, it's implicit in all our examples. Please replace with a blank line, then the html gets rendered correctly (always blank lines around `>>>`)

sgubianpm replied:
Sure ! Thanks for this.

rgommers commented on scipy/optimize/tests/test__sda.py:
```diff
@@ -0,0 +1,120 @@
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
```
This license comment is wrong, should be licensed under BSD. However best to remove all license details in individual files, except possibly (if you prefer) a simple `License: BSD`. We prefer to cover all files by the main SciPy LICENSE file.

sgubianpm replied:
Removed. Thanks, I forgot to remove from the unit tests file.

rgommers commented on scipy/optimize/tests/test__sda.py:
```diff
@@ -0,0 +1,120 @@
+"""
+Unit tests for the simulted dual annealing global optimizer
+"""
+from .._sda import sda
```
These imports should be `from scipy.optimize import sda`. We want to test the exact interface that a user will see.

sgubianpm replied:
Fixed.

sgubianpm replied:
Good point! Thanks!

rgommers commented on scipy/optimize/tests/test__sda.py:
```diff
@@ -0,0 +1,120 @@
+        values = np.zeros(dim)
+        x_step_low = vd.visiting(values, 0, self.high_temperature)
+        # Make sure that only the first component is changed
+        assert np.all(x_step_low != 0)
```
No plain asserts please (those get stripped by `python -O`). Use `np.testing.assert_` instead.

sgubianpm replied:
Changed to numpy asserts.

rgommers commented:
Hi @sgubianpm, thanks for updating your PR. I vaguely remember the details:) Looks like a reasonably clean implementation now on a quick read.

Would be helpful if you could include the benchmarking results in the PR description again.

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,568 @@
+    maxiter : int, optional
+        The maximum number of sda iterations. Increase this value if the
+        objective function is very complicated with high dimensions.
+    minimizer_kwargs : dict, optional
```
I think you should change the name of this argument to something like "local_search_options". It sounds much better and gives a better notion of what this argument is for.

sgubianpm replied:
Thanks for this. I was tempted to use the same argument name that basin hopping is offering but if you think this is more easy to read, I change it to "local_search_options"

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,568 @@
+    SDA is an implementation of the Simulated Dual Annealing. This stochastic
+    approach generalizes CSA [2]_ (Classical Simulated Annealing) and FSA (Fast
+    Simulated Annealing) to find the neighborhood of minima and introduces an
+    additional annealing process for the best solution found.
```
Please rewrite this sentence. The end of it ("to find the neighborhood of minima and introduces an additional annealing process for the best solution found.") is a little confusing.

sgubianpm replied:
Ok. Rephrased and included all the references.

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,568 @@
+
+    References
+    ----------
+    .. [1] Tsallis C (1988). "Possible generalization of Boltzmann-Gibbs
```
I think it would be nice if you mentioned all the references along the docstring text.

sgubianpm replied:
Addressed int the above comment.

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,568 @@
+        Specify `seed` for repeatable minimizations. The random numbers
+        generated with this seed only affect the visiting distribution
+        function and new coordinates generation.
+    pure_sa: boolean, optional
```
This is quite minor, but I think "no_local_search" may be a better name for this argument.

sgubianpm replied:
Ok. Changed to no_local_search

sgubianpm commented:
Thanks a lot @antonior92, I have addressed your comments and added the benchmarks results in the PR description. Any other changes you think I should do?

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,572 @@
+        return self.func(x, *self.fun_args)
+
+    def gradient(self, x):
+        g = approx_derivative(self.func_wrapper, x)
```
I just take a look at the internals of L-BFGS-B function, and it seems like the FORTRAN implementation already has an internal implementation of finite-differences. I think it would be better to use it rather than the SciPy implementation (as I have suggested at first). I think it is more elegant, since you delegate the task of approximating derivatives entirely to ``minimize``.

To use it you only need to omit the gradient as argument to ``minimize``.

sgubianpm replied:
I have made some tests and it seems that the default finite-differences calculation is less accurate than the SciPy approx_derivative (which uses 3-points by default). Therefore, I would suggest that for the default sda local minimizer, it uses the approx_derivative as you suggested first.

antonior92 replied:
Thank you very much for digging into this @sgubianpm. I think there are two points we should consider:

1. I think we should try to make the function ``sda`` as independent as possible from ``minimize``. And delegating the derivatives computation entirely to minimize seems like a good practice.
2. 2-points vs 3-point derivatives approximation is not such a simple decision. While 3-point is more accurate it also demands more functions evaluations. So it is not necessarily better.

Anyway, I don't think the internal implementation is necessarily better, but I think using it would result in a more modular and clean implementation.

sgubianpm replied:
Ok, I'll remove the custom gradient and let minimize do the job.

antonior92 replied:
Great :)

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,572 @@
+    def local_search(self, x, maxlsiter=None):
+        mres = self.minimizer(self.func_wrapper, x, **self.kwargs)
+        if not mres.success:
+            return BIG_VALUE, None
```
I think the way you just return BIG VALUE whenever the ``mres.success=false`` is not very good.

Often the local search return failure because it does not managed to achieve a solution with the specified precision within the maximum number of interactions. Nevertheless the obtained solution may still be a good solution (maybe better than the other points from your population).

I think you should replace this line by some criteria where you actually check the evaluated function at the current point. It will probably make your implementation more robust.

sgubianpm replied:
Thanks for this, agreed. I changed it so that in case it returns a failure, the original starting location and corresponding objective value is used as a returned value.

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,572 @@
+        if self.ls_max_iter < 100:
+            self.ls_max_iter = 100
+        if self.ls_max_iter > 1000:
+            self.ls_max_iter = 1000
```
I think you should rewrite this as:
```
    MAXITER_RATIO = 6
    MAXITER_MAX = 1000
    MAXITER_MIN = 100
    ...
    n = self.lower.size
    self.ls_max_iter = min(max(n*MAXITER_RATIO, MAXITER_MIN), MAXITER_MAX)

antonior92 replied:
Actually I would recommend you to try to use constants to all numbers that are arbitrary in your implementation. You have done it to some extension, but there are still many others numbers that could be replaced by constants.

This improve the readability of your code, since it helps differentiating between arbitrary constants and factors that have some strong mathematical reason to be there.

js850 replied:
Also, those values seem fairly arbitrary and potentially very dependent on the objective function.   When I first looked at this I couldn't understand why you were hardcoding these value in, but then I realized it's only used for the default minimizer if nothing else is specified.  How about you move this into another method called, e.g. setup_default_minimizer.  That will make it clear that they are only used for that purpose.

antonior92 replied:
@js850 by my personal experience, it may be too hard to explain in the documentation  and allow the user to specify all internal constants.

For instance, for the ``trust-constr`` (in preparation #7729) and ``trust-exact`` (#6919) I had a lot of internal constants, and many of those the values  were described in the original papers and worked well for me in the tests I did. In this case, I decided to keep those hard-coded (as constants), allowing the user to pass only the most important values. This way the user would have only something about 5 to 10 parameters to tune.

So I believe the best approach to be follow here is:
- Allow the user to specify parameters that have a clear meaning and large impact in the result.
- Hard code the ones that either: 1. does not make much difference in the final value; or, 2. have a good default and a not very clear meaning.

antonior92 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,572 @@
+        return (mres.fun, mres.x)
+
+
+class SDARunner(object):
```
It seems to me that you actually don't need this class at all, since you could just run the code inside ``__init__``, ``search`` and ``result`` sequentially.

It is not very critical, but I think it may make your code simpler if you just move this blocks of code to inside ``sda`` function, since you are already running them sequentially.

sgubianpm replied:
SDARunner class removed, and it actually make the code simpler.

js850 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,572 @@
+    Notes
+    -----
+    SDA is an implementation of the Simulated Dual Annealing. This stochastic
+    approach [3]_ implements the generalization of CSA (Classical Simulated
```
Is there a set of parameters to do only classical simulated annealing?  I think many people would find that useful, at the very least for comparison.  It might be nice to add an example below for that or specify how in the documentation.

sgubianpm replied:
There is no way for user to do either CSA of FSA but by using no_local_search option to True, GSA (Generalized Simulated Annealing) will be applied. Maybe this can be said in the argument docstring. I'll do that.

js850 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,572 @@
+        If `no_local_search` is set to `True`, a traditional Generalized
+        Simulated Annealing will be performed with no local search
+        strategy applied.
+
```
Is there some callback function or something or some other mechanism so people can inspect what's happening during each iteration?

sgubianpm replied:
Not really but this could be done. IMHO, this would be useful only if user is interested of the evolution of temperature of the system and when local search is performed but for the rest, the user can track coordinates and energy value in objective function directly, thoughts?

js850 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,572 @@
+    min_visit_bound = 1.e-10
+
+    def __init__(self, lb, ub, qv, rs):
+        self.qv = qv
```
I would strongly suggest that you use descriptive variable names.  I have no clue just looking at this what qv, rs, b_range, x_gauss, s_gauss are.  Same goes for SDARunner.es and many of the other SDARunner variables

sgubianpm replied:
Ok. I will rename them (except for qv, this is the equation term name, maybe, cannot really do better)

js850 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,572 @@
+        self.fun_args = None
+        lu = list(zip(*bounds))
+        self.lower = np.array(lu[0])
+        self.upper = np.array(lu[1])
```
I can't tell that self.lower and self.upper are used anywhere

sgubianpm replied:
Thanks for this. Good catch, they were used in the default gradient implementation but @antonior92 suggested to use the approx_derivative function that was in _numdiff.py, which was a good idea and I did so. It is not used anymore, I will remove it.

sgubianpm replied:
I will replace this by bounds because it is needed to pass the bounds as arguments for default minimizer settings.

js850 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,572 @@
+            self.fun_args = self.kwargs.get('args')
+
+    def func_wrapper(self, x):
+        self.nb_fun_call += 1
```
This is not a safe way to keep track of the number of function calls.  Some functions return both the func value and the gradient in one call for performance reasons.  You should probably use what the minimizer reports back after a minimiziation.  Also, you will probably want to keep track of the number of gradient calls.  Have a look how basinhopping does this

sgubianpm replied:
I'm not sure I agree with this. Either the local search which call the minimizer (user one or default), the gradient or the SDA process, all use func_wrapper. There is no call to func directly, it always goes through the wrapper and therefore counting the actual number of objective function call. Did I miss something here?

js850 commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,572 @@
+                'disp': None, 'maxls': 100, 'iprint': -1, 'gtol': 1e-06,
+                'eps': 1e-06,
+                'maxiter': self.ls_max_iter,
+                'maxcor': 10, 'maxfun': 15000
```
Where did you come up with these default values?  Why not just use the default for the L-BFGS-B minimizer?

sgubianpm replied:
These value where the ones used in our old FORTRAN implementation of GSA. I can also put the default values, it is not going to impact anything, maybe very slight changes in the benchmark, not even sure about it. I'll change it.

andyfaff commented:
Chipping in from the sidelines here - `sda` is rather cryptic. Would it be better to call it by a more descriptive name, but one that doesn't confuse ex-users of the now removed anneal code?

sgubianpm commented:
Dear All,
Thanks a lot for your time and comments. I have just commited a version addressing:
* @js850 comments about:
   - variables names (I think they are far more explicit now)
   - constant definitions for default minimizer (set_default_minimizer makes it simpler now)
* @antonior92 comments about:
   - constant definitions for default minimizer using min/max
   - Removal of the SDARunner class to make code simpler
   - Delegate the finite-differences/gradient computation to the SciPy minimize function
   - Use the default parameters for the L-BFGS-B minimize arguments
   
For @andyfaff comment about the "cryptic" sda function name. The method is called Simulated Dual Annealing, anyone with a better name (more explicit)?

Cheers,

Sylvain.

antonior92 commented:
Thank you very much for the improvements @sgubianpm. 

A posible name could be ``simulated_annealing``.

sgubianpm commented on scipy/optimize/_sda.py:
```diff
@@ -0,0 +1,544 @@
+        x_tmp = np.copy(x)
+        mres = self.minimizer(self.func_wrapper, x, **self.kwargs)
+        if not mres.success:
+            return (self.func_wrapper(x_tmp), x_tmp) 
```
Thanks @antonior92 for this, I will change it and I agree that this makes more sense. What about `dual_simulated_annealing` for the name of the function?

antonior92 commented:
My general feeling is that ``simulated_annealing`` sounds better than ``dual_simulated_annealing`` because it is a little shorter. 

I understand that you have implemented something that goes a beyond what a simple simulated annealing algorithm would do, so the more complete name make sense. But three words seems a little too long for me.

sgubianpm commented:
I have changed the name to simulated_annealing and adjusted local_search returned value in case of abnormal cases as suggested by @antonior92. Thanks a lot!

js850 commented:
@sgubianpm, @antonior92 I really don't think it's appropriate to call it simulated_annealing.  If people see a method simulated_annealing I think they would expect it to do classical simulated annealing.  I agree that a long name like dual_simulated_annealing is perhaps not optimal, but it's far better than confusing people.

antonior92 commented:
I don’t see it as much confusing. I only think it has to be well explained in the documentation.

Furthermore, the option “no_local_search” converts it to a generalized simullated anealing, right? So it does implement something in the line of a more classical algorithm as well.

js850 commented:
In addition to the local search, the VisitingDistribution is pretty novel and far from what most people would think of as classical simulated annealing.  I think the difference is big enough to warrant a novel name.  

Alternatively, you could generalize the code so that there could be some combination of input parameters that would reduce it to classical simulated annealing.  Then I think calling it simulated_annealing might be justified.

sgubianpm commented:
@antonior92, @js850, thanks for this. Just to add more insights regarding the changes compared to the old classical implementation. The version that you get when not using the local search is GSA (generalized form, a combination of CSA and FSA) which makes the visiting distribution also different as @js850 said. Where I agree with @antonior92 is that the method is still in the simulated annealing family. So I don't know, I let you guys decide what's better.

sgubianpm commented:
Dear @js850 , @andyfaff , @rgommers , @antonior92 ,
Anything in your opinion I should do next. 1.1.0 release date is approaching and I want to make sure this is getting ready. The last questionning was about the module/function name Thanks a lot for your time.
Sylvain.

andyfaff commented:
Thanks for your patience. Reviewing PR's (especially large ones) is often a
hard task because the reviewer would often like to be writing code instead.
I know from personal experience the frustration that can be experienced
when effort goes into writing code, and it gets stalled in a review stage.
However, it's still worth it in the long run.
Timely and in-depth code review is something that all OSS project's
probably fail to reach their ideals in (you can have one or the other),
your PR hasn't been forgotten, we just haven't got around to it.
Any contributions to the project (code, reviewing, documentation, bug
reporting) are always welcomed.

antonior92 commented:
Hi @sgubianpm,

Sorry for disappearing for the past months. I agree with you that we should try to make this ready this for scipy 1.1. I will try to give you some final feedback along this week.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+
+    def visit_fn(self, temperature):
+        factor1 = np.exp(np.log(temperature) / (self.visiting_param - 1.0))
+        factor2 = np.exp((4.0 - self.visiting_param) * np.log(
```
1. I think I have already said that, but I think you need to define the values such as `4.0`, `2.0`, `0.5`... as constants with more explicative names.
2. I don't like the use of the name factor1, 2, 3,... for variables, maybe you should come up with more meaningful names.

sgubianpm replied:
Hi @antonior92, thanks for the comments. This function, variable names and constants are provided by Constantino Tsallis in the paper (Tsallis C, Stariolo DA (1996). "Generalized Simulated Annealing." Physica A, 233, 395-406.). There is no particular meaning for it and it is the formula expression (look at page 405 of the paper).

antonior92 replied:
Ok, so maybe include a comment citing the formula from the original paper. Something like:

```
# Formula X from p. 405 of reference [2]
```

sgubianpm replied:
Commented as such. Thanks!

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+    chain, the class implements the strategy for generating new location
+    changes.
+    """
+    tail_limit = 1.e8
```
I think constants should be named using all UPPER CASE by pep 8 convention.

sgubianpm replied:
Adjusted.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+        }
+        self.kwargs['bounds'] = list(zip(self.lower, self.upper))
+
+    def func_wrapper(self, x):
```
I think you actually do not need this function wrapper.

Can you not use the function count from minimize? Besides that minimize also supports passing arguments for the function.

sgubianpm replied:
Done using nfev from OptimizeResult returned by minimize for counting objective function calls during the local search. fun_wrapper removed.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+        ``f(x, *args)``, where ``x`` is the argument in the form of a 1-D array
+        and ``args`` is a  tuple of any additional fixed parameters needed to
+        completely specify the function.
+    x0 : ndarray
```
I have never studied simulated_annealing algorithm  but I have always thought it to be a population algorithm. So just a question, `x0` is a single initial condition point or a list of points?

This is not very clear from the documentation.

Besides that, I like it very much when the documentation specify the dimensions of the ndarray. Something like:
```
x0 : ndarray, shape (n, m)
    ...
bounds : sequence, shape (n,)
    ...
```

sgubianpm replied:
x0 is the single optional initial coordinates for the algorithm to start. It can be used as a initial guess. I'll make this more clear in the documentation and specify dimensions. Thanks for this.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+    maxiter : int, optional
+        The maximum number of simulated_annealing  iterations. Increase this
+        value if the objective function is very complicated with high
+        dimensions.
```
I don't like the comment "Increase this value if the objective function is very complicated with high dimensions.", I think it is an oversimplification and we are probably better without it in the documentation.

It is worth including a phrase giving the default. Something like ``Default is 1000.``

antonior92 replied:
This number does not count for the number of local search iterations right?

Please try to clarify this.

sgubianpm replied:
Adjusted accordingly. This is the overall algorithm number of iteration, it may for some iterations not use the local search, so it is not related to local search number of iterations.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+                its derivatives (Jacobian, Hessian).
+    initial_temp : float, optional
+        The initial temperature, use higher values to facilitates a wider
+        search of the energy landscape, allowing simulated_annealing  to escape
```
double space between ``simulated_annealing`` and ``to escape``.

antonior92 replied:
Include a phrase describing the default too.

sgubianpm replied:
Thanks. Done.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+        no_local_search=False):
+    """
+    Find the global minimum of a function using the Simulated Dual Annealing
+    algorithm
```
Include punctuation in the end of the phrase.

sgubianpm replied:
Done. Thanks.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+        Specify `seed` for repeatable minimizations. The random numbers
+        generated with this seed only affect the visiting distribution
+        function and new coordinates generation.
+    no_local_search : boolean, optional
```
I think it should be ``bool`` instead of ``boolean``.

sgubianpm replied:
Ok. Fixed.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+        algorithm is in the middle of a local search, this number will be
+        exceeded, the algorithm will stop just after the local search is
+        done.
+    seed : int or `np.random.RandomState`, optional
```
I prefer:
`` seed : {int, `np.random.RandomState`}, optional``

antonior92 replied:
Could not the user just set:
```
np.random.seed()
```
in order to get something reproducible?

Do we actually need a parameter for that?

sgubianpm replied:
Thanks. Doc adjusted. I think we do need a parameter for that and using `RandomState` was a better option, this was discussed with @js850 and @andyfaff .

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+        The initial temperature, use higher values to facilitates a wider
+        search of the energy landscape, allowing simulated_annealing  to escape
+        local minima that it is trapped in.
+    visit : float, optional
```
give the default

sgubianpm replied:
Changed accordingly.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+    visit : float, optional
+        Parameter for visiting distribution. Higher values give the visiting
+        distribution a heavier tail, this makes the algorithm jump to a more
+        distant region. The value range is (0, 3]
```
Add punctuation.

antonior92 replied:
``The value range is (0, 3].``

sgubianpm replied:
Done

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+    accept : float, optional
+        Parameter for acceptance distribution. It is used to control the
+        probability of acceptance. The lower the acceptance parameter, the
+        smaller the probability of acceptance. It has to be any negative value.
```
The phrase ``It has to be any negative value.`` is a little strange. Reformulate.

sgubianpm replied:
Fixed.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+    implementation was done in C++ for the R language [5]_ and has been
+    benchmarked [6]_. Extensive benchmarks for the Python version have also
+    been performed against several global optimization methods [7]_. SDA
+    introduces an advanced dual annealing method to refine the solution found
```
I do not like the phrase:

``A first  implementation was done in C++ for the R language [5]_ and has been benchmarked [6]_. Extensive benchmarks for the Python version have also  been performed against several global optimization methods [7]_.``

 it gives very few information that is actually useful to someone using your code. Maybe reduce it to something like:
``Alternative implementations of this same algorithm are described in [5]_  and benchmarks are presented in [6]_ and [7]_.``

sgubianpm replied:
Adjusted.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+        Statistical Software, 60(6), 1 - 45.
+        http://dx.doi.org/10.18637/jss.v060.i06
+    .. [7] Xiang Y, Gubian S, Martin F. Simulated dual annealing method for
+        efficient global optimization. Forthcoming.
```
I do not like the idea of citing a forthcoming paper in the documentation.

I suggest:
1. Put the paper on ArXiV and cite the pre-print; or
2. Remove this citation.

sgubianpm replied:
Ok. Reference removed.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+
+    .. versionadded:: 1.1.0
+
+    References
```
Please put all the references in a standard format.

sgubianpm replied:
Reformatted.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+        Annealing Algorithm and Its Application to the Thomson Model."
+        Physics Letters A, 233, 216-220.
+    .. [4] Xiang Y, Gong XG (2000a). "Efficiency of Generalized Simulated
+        Annealing." PHYSICAL REVIEW E, 62, 4473.
```
``Physical Review E``

sgubianpm replied:
Adjusted.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+    >>> up = [5.12] * 10
+    >>> ret = simulated_annealing(func, None, bounds=list(zip(lw, up)))
+    >>> print("global minimum: xmin = {0}, f(xmin) = {1}".format(
+    ...    ret.x, ret.fun))
```
Include the result in the example.

sgubianpm replied:
Example results added.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+                               energy_state)
+
+    # Run the search lopp 
+
```
Why do you break the line here? I do not see any need for it

sgubianpm replied:
Removed.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,559 @@
+    bounds : sequence
+        Bounds for variables.  ``(min, max)`` pairs for each element in ``x``,
+        defining the lower and upper bounds for the optimizing argument of
+        `func`. It is required to have ``len(bounds) == len(x)``.
```
Please use a more compact explanation, such as:
```
x0 : ndarray, shape (n, m)
    ...
bounds : sequence, shape (n,)
    ...
```

sgubianpm replied:
Compacted.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+    approach [3]_ implements the generalization of CSA (Classical Simulated
+    Annealing) and FSA (Fast Simulated Annealing) [1]_ [2]_ coupled to a
+    strategy for applying a local search on accepted locations [4]_.
+    Alternative implementation of this same algorithm is described in [5]_ and
```
I think it should be ``An alternative implementation...``

sgubianpm replied:
Fixed, thanks a lot.

antonior92 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,558 @@
+            while enter or (self.s_gauss <= 0 or self.s_gauss >= 1):
+                enter = False
+                sample1 = self.rand_state.random_sample()
+                self.x_gauss = sample1 * 2.0 - 1.0
```
If these numbers are from a paper as well. I think you can add a reference here too.

antonior92 commented:
I added one last comment for you. Also, you have to fix the continuous integration issues. After that I think the code will be in good shape and could be merged.

@rgommers and @js850 what do you think? Any additional comments?

js850 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+# Simulated Dual Annealing implementation.
+# Copyright (c) 2018 Sylvain Gubian <sylvain.gubian@pmi.com>,
```
Should this copyright statement be here?  (I'm just asking, I have no clue).

sgubianpm replied:
I thought it was ok as I referred to an [open source guide](https://opensource.guide/) and [this one](http://mg.readthedocs.io/python.html) but if you think this should be differently, I would be happy to adjust.

js850 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+                else:
+                    pqv = np.exp(np.log(pqv_temp) / (
+                        1. - self.acceptance_param))
+                if r <= pqv:
```
I think it would be clearer if you separate out this acceptance criterion into a separate method.

js850 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+                if self.not_improved_idx >= self.not_improved_max_idx:
+                    if j == 0 or self.state.current_energy < self.emin:
+                        self.emin = self.state.current_energy
+                        self.xmin = np.copy(self.state.current_location)
```
What is the purpose of this block?  if local_search is disabled emin will only be updated here.  Does that make sense?

js850 commented on scipy/optimize/tests/test__simulated_annealing.py:
```diff
@@ -0,0 +1,115 @@
+        assert_array_less(np.min(values), 1e-10)
+        assert_array_less(1e+10, np.max(values))
+
+    def test__reset(self):
```
some methods have 2 underscores, some have 1.  Please standardize

sgubianpm replied:
Done

js850 commented on scipy/optimize/tests/test__simulated_annealing.py:
```diff
@@ -0,0 +1,115 @@
+    def test_bounds_integrity(self):
+        wrong_bounds = [(-5.12, 5.12), (1, 0), (5.12, 5.12)]
+        assert_raises(ValueError, simulated_annealing, self.func, None,
+            wrong_bounds)
```
Do you test correctness with local_search disabled?

sgubianpm replied:
Added tests:
- `test_low_dim_no_ls`
- `test_high_dim_no_ls`
- `test_nb_fun_call`
- `test_nb_fun_call_no_ls`

js850 commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+                3.0 - self.visiting_param))
+        return x / den
+
+    def gaussian_fn(self, axis):
```
What is this method doing?  Drawing a random variable from a gaussian distribution?  Doesn't numpy already have a method for that?

If it's doing something else, please add an explanation

sgubianpm replied:
You are right, the numpy version also implements the Marsaglia polar method for it, which is the same thing (except numpy uses a C extention). Let's use the numpy version then.

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+                if pqv_temp < 0.:
+                    pqv = 0.
+                else:
+                    pqv = np.exp(np.log(pqv_temp) / (
```
is this numerically stable? Can acceptance_param ever equal 1 --> divide by zero errors.

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+                    self.state.current_location = np.copy(x_visit)
+                    self.xmin = np.copy(self.state.current_location)
+
+                # No improvement since long time
```
since --> for a

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+        if 'args' in self.kwargs:
+            self.fun_args = self.kwargs.get('args')
+        
+    def set_default_minimizer(self):
```
missing a docstring

andyfaff replied:
is it necessary to have this method?  Can't you set up the object with these as the default, and overwrite them if the options are in `kwargs`?

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+    MAXITER_MIN = 100
+    MAXITER_MAX = 1000
+
+    def __init__(self, bounds, func, **kwargs):
```
missing a docstring on how to initialise the object

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+        x_tmp = np.copy(x)
+        fun_temp = self.func(x, *self.fun_args)
+        mres = self.minimizer(self.func, x, *self.fun_args, **self.kwargs)
+        self.nb_fun_call += mres.nfev 
```
what happens if the minimizer doesn't support bounds.

sgubianpm replied:
Solved by addressing the comment above.
The minimizer can use unbounded space but the implementation checks that the results is in bounds otherwise, it is rejected.

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+        completely specify the function.
+    x0 : ndarray, shape(n,)
+        A single initial starting point coordinates. If ``None`` is provided,
+        initial coordinates are automatically generated.
```
how are they automatically generated?

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+        Extra keyword arguments to be passed to the local minimizer
+            ``scipy.optimize.minimize()`` Some important options could be:
+            method : str
+                The minimization method (e.g. ``"L-BFGS-B"``)
```
does the method have to use bounds?

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+            args : tuple
+                Extra arguments passed to the objective function (``func``) and
+                its derivatives (Jacobian, Hessian).
+    initial_temp : float, optional
```
what is a sensible range for initial_temp? Are there definite bounds that shouldn't be exceeded?

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+        Parameter for visiting distribution. Higher values give the visiting
+        distribution a heavier tail, this makes the algorithm jump to a more
+        distant region. The value range is (0, 3]. Default value is 2.62. 
+    accept : float, optional
```
what is a sensible range for `accept`? Are there definite bounds that shouldn't be exceeded?

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+    >>> lw = [-5.12] * 10
+    >>> up = [5.12] * 10
+    >>> ret = simulated_annealing(func, None, bounds=list(zip(lw, up)))
+    >>> #print("global minimum: xmin = {0}, f(xmin) = {1:.6f}".format(
```
Why is the `print` commented out?

sgubianpm replied:
It was creating an error Travis CI complaining for the output. I will check this and restore the example.

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+
+def simulated_annealing(func, x0, bounds, maxiter=1000, local_search_options={},
+        initial_temp=5230., visit=2.62, accept=-5.0, maxfun=1e7, seed=None,
+        no_local_search=False):
```
missing a `callback` mechanism, a callable that is called at the end of every iteration with the best solution so far. See e.g. L-BFGS-B for how to implement.

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+            method : str
+                The minimization method (e.g. ``"L-BFGS-B"``)
+            args : tuple
+                Extra arguments passed to the objective function (``func``) and
```
this should be a keyword argument to `simulated_annealing`, not a part of the `local_search_options`

andyfaff commented on scipy/optimize/tests/test__simulated_annealing.py:
```diff
@@ -0,0 +1,115 @@
+# Simulated dual annealing unit tests implementation.
```
The test suite does not have comprehensive coverage , with various tests missing. E.g.

- testing that maxfun works
- testing that maxiter works
- testing that the args tuple is passed correctly
- testing correct and incorrect bounds
- `VisitingDistribution` and `MarkovChain`, `EnergyState` are missing comprehensive testing.
- etc

As an example, I'd hope that something approximating the depth of `optimize.tests.test__differential_evolution` could be approached.

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+    markov_chain = MarkovChain(accept, visit_dist, obj_fun_wrapper, rand_state,
+                               energy_state)
+
+    # Run the search lopp 
```
lopp-->loop

andyfaff commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+    def local_search(self, x):
+        x_tmp = np.copy(x)
+        fun_temp = self.func(x, *self.fun_args)
+        mres = self.minimizer(self.func, x, *self.fun_args, **self.kwargs)
```
`args` is passed into `minimize` as a tuple, but you're unpacking the tuple here. Shouldn't it be passed in like:

```
self.minimizer(self.func, x, args=self.fun_args, **self.kwargs)
```
?

andyfaff replied:
How does simulated_annealing deal with a local minimizer failure?

sgubianpm commented:
Dear @andyfaff , @js850 , @antonior92 
Thanks for the new incoming comments ;-)
I'll do my best to address them asap...

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+            return self.square_root_gaussian * self.x_gaussian_sample
+
+
+class EnergyState():
```
Sure, thanks.

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+            obj_fun_wrapper.nb_fun_call += 1
+            if self.current_energy is None:
+                raise ValueError('Objective function is returning None')
+            if (self.current_energy >= BIG_VALUE or
```
Yes, thanks a lot. Replaced and BIG_VALUE removed.

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+                    init_error = False
+                    message = (
+                        'Stopping algorithm because function '
+                        'create NaN or (+/-) inifinity values even with '
```
Fixed and now ok with comment above.

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+    
+    MAXITER_RATIO = 6
+    MAXITER_MIN = 100
+    MAXITER_MAX = 1000
```
I prefixed it with LS_ so that it is more clear that it is related to local search.

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+
+    def local_search(self, x):
+        x_tmp = np.copy(x)
+        fun_temp = self.func(x, *self.fun_args)
```
Thanks for that, the test added in the comment below was demonstrating that the increment was missing here (and 3 calls were missing over 4093). Anyway, this extra fun call  can be avoided by passing the starting energy value to local_search function. Fixed.

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+        self.emin = state.current_energy
+        self.xmin = np.array(state.current_location)
+        # Global optimizer state
+        self.state = state
```
Done.

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+            self.obj_fun_wrapper.nb_fun_call += 1
+            if e < self.state.current_energy:
+                #  print('Better energy: {0}'.format(e))
+                # We have got a better ernergy value
```
Fixed

sgubianpm commented on benchmarks/benchmarks/optimize.py:
```diff
@@ -207,12 +207,14 @@ def bench_run(self, x0, methods=None, **minimizer_kwargs):
 
         if methods is None:
-            methods = ["COBYLA", 'Powell',
+            methods = ["COBYLA", 'Powell', 'nelder-mead',
```
I will do a new PR with all addressed comments as this one went for a long time, and will also allow to clear this up.

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+                x_visit[index] += self.MIN_VISIT_BOUND
+        return x_visit
+
+    def visit_fn(self, temperature):
```
Docstring added that gives the publication reference for.

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+        self.square_gaussian = 0
+        self.square_root_gaussian = None
+
+    def visiting(self, x, step, temperature):
```
Docstring added. Thanks for that.

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+    Cauchy-Lorentz distribution. Depending on the steps within the Markov
+    chain, the class implements the strategy for generating new location
+    changes.
+    """
```
Docstring added with parameters description

sgubianpm commented on scipy/optimize/_simulated_annealing.py:
```diff
@@ -0,0 +1,557 @@
+        self.temperature_step = temperature / float(step + 1)
+        self.not_improved_idx += 1
+        for j in range(self.state.current_location.size * 2):
+            if j == 0:
```
Done.

sgubianpm commented:
I close this one as it is very long and old now. New one with comments addressed is #9306.
